x/mongo/driver/topology/server.go

```diff
@@ -540,15 +540,18 @@ func (s *Server) heartbeat(conn *connection) (description.Server, *connection) {
 			if err == nil {
 				tmpDesc := op.Result(s.address)
 				descPtr = &tmpDesc
+			} else {
+				// close the connection here rather than in the error check below to avoid calling Close on a net.Conn
+				// that wasn't successfully created
+				if conn != nil && conn.nc != nil {
+					_ = conn.nc.Close()
+				}
 			}
 		}
 
 		// we do a retry if the server is connected, if succeed return new server desc (see below)
 		if err != nil {
 			saved = err
-			if conn != nil && conn.nc != nil {
```

Comment:
Is there no other way to arrive here where an `err != nil` other than from the conditional statement above? For instance, the code on 526 `err = conn.wait()` seems like it would fall into this bucket, and now we aren't closing the conn any longer.

Reply:
`conn.wait` is the method that actually creates the network connection and performs the handshake. If it fails, it closes the connection before returning. This was done because a failure could happen at different points (e.g. dialing vs handshaking) and it's the only function that knows exactly what resources need to be cleaned up.

Reply:
In looking through the conn.connect method (of which wait just returns the error from connect if one existed), it appears as though we may leak an open connection if the TLS handshake fails. It's actually worse than that because we actually lose the underlying connection as well because it was overwritten with nil, but never actually closed. In this case, we'd get a nil pointer issue.

Reply:
I see what you're saying. I think my new commit fixes this.